src/collapse/xctrace.rs

```diff
@@ -0,0 +1,455 @@
+    }
+}
+
+fn get_id_from_attributes(attributes: &Attributes) -> io::Result<u64> {
+    let id = attributes
+        .clone()
+        .filter_map(|x| x.ok())
+        .find_map(|x| (x.key.into_inner() == ID).then_some(x.value));
+    let id = match id {
+        Some(x) => x,
+        None => return invalid_data_error!("No id found in attributes"),
+    };
+    let id = String::from_utf8_lossy(&id);
+    match id.parse() {
+        Ok(x) => Ok(x),
+        Err(e) => invalid_data_error!("Unrecognized id: {}: {:?}", id, e),
+    }
+}
+
+fn get_name_from_attributes(attributes: &Attributes) -> io::Result<Vec<u8>> {
+    let name = attributes
+        .clone()
+        .filter_map(|x| x.ok())
+        .find_map(|x| (x.key.into_inner() == NAME).then_some(x.value));
+    match name {
+        Some(x) => Ok(unescape_xctrace_text(x)),
+        None => invalid_data_error!("No name(symbol) found in attributes"),
+    }
+}
+
+/// Extract necessary info from attributes for constructing backtrace.
+fn attributes_to_backtrace(attributes: &Attributes) -> io::Result<u64> {
+    get_id_from_attributes(attributes)
+}
+
+/// Extract necessary info from attributes for constructing frame.
+fn attributes_to_frame(attributes: &Attributes) -> io::Result<(u64, Vec<u8>)> {
+    let id = get_id_from_attributes(attributes)?;
+    let name = get_name_from_attributes(attributes)?;
+    Ok((id, name))
+}
+
+/// A stack collapser for the output of `xctrace export`.
+#[derive(Default)]
+pub struct Folder;
+
+impl Collapse for Folder {
+    fn collapse<R, W>(&mut self, reader: R, writer: W) -> std::io::Result<()>
+    where
+        R: std::io::BufRead,
+        W: std::io::Write,
+    {
+        let mut reader = Reader::from_reader(reader);
+        self.collapse_inner(&mut reader, writer)
+    }
+
+    fn is_applicable(&mut self, input: &str) -> Option<bool> {
+        let mut input = input.as_bytes();
+        let mut line = String::new();
+        loop {
+            if let Ok(n) = input.read_line(&mut line) {
+                if n == 0 {
+                    break;
+                }
+            } else {
+                return Some(false);
+            }
+
+            let trimmed = line.trim();
+            if !trimmed.is_empty() {
+                return Some(trimmed.contains(r#"<?xml version="1.0"?>"#));
+            }
+            line.clear();
+        }
+        None
+    }
+}
+
+impl Folder {
+    fn collapse_inner<R, W>(&mut self, reader: &mut Reader<R>, writer: W) -> io::Result<()>
+    where
+        R: std::io::BufRead,
+        W: std::io::Write,
+    {
+        let mut context = CollapseContext::default();
+        let mut buf = Vec::new();
+        let nodes = loop {
+            let event = match reader.read_event_into(&mut buf) {
+                Ok(x) => x,
+                Err(e) => return invalid_data_error!("Read xml event failed: {:?}", e),
+            };
+            match event {
```

Comment:
It would be really helpful to have a comment here that shows some example XML we're trying to parse to get a sense of the layout.